### PR TITLE
DHC code

### PR DESCRIPTION
**Pull request by langtaibai**
1. For partition, use temp file to mount , and use native call function to deal with disk/partition instead of hardcoding the disk/partition.
2. For scalable, use existing parameters(rec_type) to relate  DHC flow instead of creating a new parameter, and use platform list config file to update new platform instead of change dell-recovery code.
3. For copy file, use white_tree function to copy file instead of using os.system , and use one function call to copies of multiple directories to avoid code duplicate.

### CODE REVIEW

**Comment by superm1** [comment]
There are conflicts listed, please make sure that you have rebased your code on master for this merge request to be able to work.

File: Dell/recovery_common.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
 gi.require_version('UDisks', '2.0')
 from gi.repository import GLib, UDisks
 import os
-import shutil
+#import shutil
```
Comment (superm1) [inline Dell/recovery_common.py]: Don't comment out an import, if it's not needed - please remove it.

File: Dell/recovery_common.py [hunk 2]
```diff
@@ -655,12 +691,11 @@ def parse(fname):
                     arch, package, modaliases = parse(os.path.join(repo, fname))
                     if not modaliases and (arch == "all" or arch == my_arch):
                         to_install.append(package)
-
```
Comment (superm1) [inline Dell/recovery_common.py]: for this and others, don't change whitespace in the commit.  If whitespace was wrong previously, please have a separate follow up to fix it from your actual code changes commits.

File: Dell/recovery_common.py [hunk 3]
```diff
@@ -655,12 +691,11 @@ def parse(fname):
...
     return to_install
 
 def create_grub_entries(target_dir='/target', rec_type='hdd'):
     '''Create GRUB entry for dell-recovery during ubiquity installation'''
-    env = os.environ
+    #env = os.environ
```
Comment (superm1) [inline Dell/recovery_common.py]: if i'm not mistaken, it was important to have the environment set here so that the language would be applied as necessary for any translations that do exist.

File: codestyle.txt [hunk 4]
```diff
@@ -0,0 +1,1177 @@
+./setup.py:24:49: W291 trailing whitespace
```
Comment (superm1) [inline codestyle.txt]: why was this file added?

File: debian/changelog [hunk 5]
```diff
@@ -1,4 +1,4 @@
-dell-recovery (1.61) UNRELEASED; urgency=medium
```
Comment (superm1) [inline debian/changelog]: I think when you rebase on master, you shouldn't be changing this at all.

File: debian/changelog [hunk 6]
```diff
@@ -15,7 +15,7 @@ dell-recovery (1.61) UNRELEASED; urgency=medium
   * add lvm2 and cryptsetup to recommends so they won't be removed in
     encrypted rootfs use case.
 
- -- Yuan-Chen Cheng <yc.cheng@canonical.com>  Wed, 23 Jan 2019 10:43:06 +0800
+ -- Shih-Yuan Lee (FourDollars) <sylee@canonical.com>  Thu, 24 Jan 2019 11:25:03 +0800
```
Comment (superm1) [inline debian/changelog]: also shouldn't be changing this.

File: debian/changelog [hunk 7]
```diff
@@ -25,6 +25,7 @@ dell-recovery (1.60) disco; urgency=medium
   * Reduce some PyGIWarning messages.
   * Use the regional mirror site instead of the global mirror site.
   * Renew .disk/ubuntu_dist_channel for ubuntu-report.
+  * Support the installation on Intel RSTe SATA RAID.
```
Comment (superm1) [inline debian/changelog]: also shouldn't be changing this

File: debian/control [hunk 8]
```diff
@@ -5,8 +5,8 @@ Maintainer: Mario Limonciello <superm1@ubuntu.com>
 XSBC-Original-Maintainer: Mario Limonciello <mario_limonciello@dell.com>
 Build-Depends: debhelper (>= 11), python3-all, python3-distutils-extra, po-debconf, pyflakes, lsb-release, gnu-efi (>= 3.0.2)
 X-Python3-Version: >= 3.2
-Homepage: http://linux.dell.com/
-Vcs-Git: http://github.com/superm1/dell-recovery
+Homepage: http://github.com/dell/dell-recovery
+Vcs-Git: https://github.com/dell/dell-recovery/tree/bionic
```
Comment (superm1) [inline debian/control]: bionic?  I thought this was on master.

File: debian/dell-recovery.templates [hunk 9]
```diff
@@ -64,6 +64,30 @@ Template: ubiquity/text/hdd_warning_label
 Type: text
 _Description: WARNING: All personal files and changes will be lost.
 
+Template: ubiquity/text/dhc_hdd_recovery
+Type: text
+_Description: Restore Linux OS partitions with DHC
```
Comment (superm1) [inline debian/dell-recovery.templates]: This is a customer facing string, right?  No one will know what DHC means.  I think you need to expand the acronym.

File: debian/dell-recovery.templates [hunk 10]
```diff
@@ -78,11 +102,15 @@ _Description: WARNING: All personal files and changes will be lost.
 
 Template: ubiquity/text/interactive_recovery
 Type: text
-_Description: Restore only Linux OS partition.
+_Description: Customize installation.
```
Comment (superm1) [inline debian/dell-recovery.templates]: Why did you change this?

**Comment by superm1** [comment]
@langtaibai I didn't review everything, but I think there are some core problems with the commits not being based on top of the proper branch.  Please make sure it's on top of master.  This shouldn't be 26 commits, it should be at most 5 commits.  There shouldn't be any merge commits.  So do an interactive rebase onto master to squash and drop relevant commits and then force push to your branch.

It should be something like this:
```
# git pull origin master
# git rebase -i origin/master
```

File: Dell/recovery_gtk.py [hunk 11]
```diff
@@ -136,11 +139,24 @@ def top_button_clicked(self, widget):
...
-                                                  False)
-                    proc = subprocess.Popen(["gnome-session-quit", "--reboot"])
+                                                  True)
+                    #proc = subprocess.Popen(["gnome-session-quit", "--reboot"])
+                    self.destroy()
+                except dbus.DBusException as msg:
+                    self.dbus_exception_handler(msg)
+                       
+            #Restore DHC Button
+            else:
```
Comment (superm1) [inline Dell/recovery_gtk.py]: So does that mean that this button will be present at all times on all machines?

File: Dell/recovery_gtk.py [hunk 12]
```diff
@@ -136,11 +139,24 @@ def top_button_clicked(self, widget):
...
+                    self.dbus_exception_handler(msg)
+                       
+            #Restore DHC Button
+            else:   
+                try:
+                    dbus_sync_call_signal_wrapper(self.backend(),
+                                                  "enable_boot_to_restore_dhc",
+                                                  {},
+                                                  True)
+                    #proc = subprocess.Popen(["gnome-session-quit", "--reboot"])
```
Comment (superm1) [inline Dell/recovery_gtk.py]: Don't include commented out lines in the commits.  I suspect you actually do want this line still.

File: Dell/recovery_xml.py [hunk 13]
```diff
@@ -32,7 +32,7 @@
     text_type = str
     binary_type = bytes
 else:
-    text_type = unicode
+#    text_type = unicode
```
Comment (superm1) [inline Dell/recovery_xml.py]: similarly, don't comment out lines

File: ubiquity/dell-bootstrap.py [hunk 14]
```diff
@@ -112,7 +122,8 @@ def __init__(self, controller, *args, **kwargs):
         self.efi = False
         with misc.raised_privileges():
             self.genuine = magic.check_vendor()
-
+        with misc.raised_privileges():
```
Comment (superm1) [inline ubiquity/dell-bootstrap.py]: Does this actually need elevated privs?  I thought that `/sys/class/dmi/id/product_sku` can be read by anyone by default.

File: ubiquity/dell-bootstrap.py [hunk 15]
```diff
@@ -1023,7 +1100,25 @@ def build_rp(self, cushion=600):
                 if item.startswith('ID_FS_UUID'):
                     uuid = item.split('=')[1]
                     break
-
+        
+        #if select DHC, will copy DHC file for silent install
+        with misc.raised_privileges():
+            if self.rec_type=='dhc':
+                if os.path.exists(os.path.join("/mnt", "preseed", "dell-recovery.seed")):
+                    os.remove(os.path.join("/mnt", "preseed", "dell-recovery.seed"))
```
Comment (superm1) [inline ubiquity/dell-bootstrap.py]: Can't this just be excluded in the whitetree/blacktree functions?

File: ubiquity/dell-bootstrap.py [hunk 16]
```diff
@@ -1023,7 +1100,25 @@ def build_rp(self, cushion=600):
...
+                if not os.path.exists(os.path.join("/mnt", "scripts")):
+                    os.makedirs(os.path.join("/mnt", "scripts"))
+                dirs=["casper", "preseed", "scripts"]
+                for dir_tmp in dirs:
+                    mnt_dir=os.path.join("/mnt", "DHC", dir_tmp)
+                    assembly_dir=os.path.join("/mnt", dir_tmp)
+                    magic.white_tree("copy",re.compile(''), mnt_dir, assembly_dir)
+        #There is a script in platform_script that will copy the platform list file to the /usr/share/dell/DHC/ dir, so during recovery you can read the platform list. The way to do this is to  avoid recovery code change when add new platform, because you only need to update the file.
```
Comment (superm1) [inline ubiquity/dell-bootstrap.py]: As in it's not owned by a package?  Then it shouldn't be put into `/usr/`.  It should be going into more likely something like `/var/lib/dell-recovery`

File: Dell/recovery_backend.py [hunk 17]
```diff
@@ -777,6 +777,31 @@ def validate_driver_package(self, package, sender=None, conn=None):
...
+        in_signature = 'b', out_signature = '', sender_keyword = 'sender',
+        connection_keyword = 'conn')
+    #The funcution is used to recovery DHC
+    def enable_boot_to_restore_dhc(self, reboot, sender=None, conn=None):
+        """Enables the default one-time boot option to be recovery"""
+        self._reset_timeout()
+        recovery=find_partition().decode('utf-8')
+        mntdir=self.request_mount(recovery, 'rw', sender,conn)
+        #Create dhc_flag to provide judgment for recovery DHC to delete copied dhc file, and dhc_flag will be deleted after recovery DHC.  
+        os.makedirs(os.path.join(mntdir, "dhc_flag"))
```
Comment (superm1) [inline Dell/recovery_backend.py]: I don't see this used anywhere else in the PR  What's it used by?

File: Dell/recovery_backend.py [hunk 18]
```diff
@@ -777,6 +777,31 @@ def validate_driver_package(self, package, sender=None, conn=None):
...
+        #Delete seedfile for copy DHC seed file
+        if os.path.exists(os.path.join(mntdir, "preseed", "dell-recovery.seed")):
+            os.remove(os.path.join(mntdir,"preseed", "dell-recovery.seed"))
+        if not os.path.exists(os.path.join(mntdir, "scripts")):
+            os.makedir(os.path.join(mntdir, "scripts"))
+        #copy DHC file to corresponding dir for recovery DHC
+        dirs=["casper", "preseed", "scripts"]
```
Comment (superm1) [inline Dell/recovery_backend.py]: So if i'm following this correctly - you're duplicating everything from `/mnt/dhc/*` to `/mnt/`, right?

Why not instead just reference contained items in place?  It seems like a waste of space on the recovery partition to have to duplicate all of these.

File: Dell/recovery_common.py [hunk 19]
```diff
@@ -153,6 +153,45 @@ def check_family(test):
...
+                return False
+            lines=[line.rstrip('\n') for line in open(plat_conf)]
+            for i in range(len(lines)):
+                if lines[i] in str(value):
+                    return True
+    return False
+    
+def check_recovery_dhc_id():
+    """The function is used to detect machine's ID for recovery DHC flow,"""
+    path = '/sys/class/dmi/id/modalias'
```
Comment (superm1) [inline Dell/recovery_common.py]: I'm a bit confused - why isn't this the same as install-id flow?  Don't you want the same stuff offered for recovery or install on the same machines?

File: Dell/recovery_backend.py (hunk 17 above)
Comment (langtaibai) [reply Dell/recovery_backend.py]: This used to switch recovery Ubuntu and recovery DHC.  After recovery DHC, the scripts and seed files we copied about DHC  are still kept in the corresponding directory. If  next recovery Ubuntu at this time, then the flow of recovery DHC is still the same.  So create a script in DHC/scripts/chroot-scripts/fish  to Check if there is this flag, and if have, delete the file copy during recovery DHC.

File: Dell/recovery_backend.py (hunk 18 above)
Comment (langtaibai) [reply Dell/recovery_backend.py]: Only the scripts and seed files written for DHC are copied. For DHC package and slides, we will not copy them, but read the corresponding path through the script.

File: Dell/recovery_common.py (hunk 1 above)
Comment (langtaibai) [reply Dell/recovery_common.py]: Thanks, will remove it.

File: Dell/recovery_common.py (hunk 19 above)
Comment (langtaibai) [reply Dell/recovery_common.py]: Because the kernel version currently used for 7070 is 4.15, it does not include product_sku. If want to unify, maybe  all read modalias file

File: Dell/recovery_common.py (hunk 2 above)
Comment (langtaibai) [reply Dell/recovery_common.py]: Got it, thanks.

File: Dell/recovery_common.py (hunk 3 above)
Comment (langtaibai) [reply Dell/recovery_common.py]: Yes, don't need to change.
Thanks.

File: Dell/recovery_gtk.py (hunk 11 above)
Comment (langtaibai) [reply Dell/recovery_gtk.py]: This button is only available on platforms which need recovery DHC and recovery normal Ubuntu.

File: Dell/recovery_backend.py (hunk 18 above)
Comment (superm1) [reply Dell/recovery_backend.py]: Where possible, it's better to remove state from the equation.  Go off the presence of a single flag file to figure out which files to load rather than needing to copy pieces around.  It's much more difficult to debug problems that only crop up after a recovery or a recovery from a recovery when items have been copied around and the state isn't guaranteed identical.

You already are creating a `dhc_flag`, I would propose look for that and when it's there search for additional locations for these items.

File: Dell/recovery_backend.py (hunk 17 above)
Comment (superm1) [reply Dell/recovery_backend.py]: Just as a point of preference, would you make it `.dhc_flag` instead of `dhc_flag`?  Then it's hidden by default in the filesystem when you use something like `ls`.

File: ubiquity/dell-bootstrap.py (hunk 16 above)
Comment (langtaibai) [reply ubiquity/dell-bootstrap.py]: It's owned by DHC package, when launching the dell recovery tool, need to read the platform ID file to determine whether to display the "recovery DHC" option. When just launched, I don’t know how to give root permission to mount the Recovery partition, so I copied this file to a directory during installation, and read this directory during recovery.  So you mean using the directory /var/lib/dell-recovery

Comment (superm1) [reply ubiquity/dell-bootstrap.py]: Yes, either a state file should be stored in the stateful directory (`/var/lib`) or the cache directory (`/var/cache`).  If you need it to persist across a reinstall, you can create something in EFI NVRAM (`/sys/firmware/efi/efivars`)

File: Dell/recovery_backend.py (hunk 18 above)
Comment (langtaibai) [reply Dell/recovery_backend.py]: Yes, If using dhc_flag can search additional location, that's fine. If not,  is it possible to use the rename method, like efi directory.
